### PR TITLE
Add piston extend and retract events

### PR DESCRIPTION
**Pull request by Selim042**
Comes directly from #2240, just updated.

### CODE REVIEW

File: src/main/java/net/minecraftforge/event/world/PistonEvent.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    {
+        public PistonExtendEvent(World world, BlockPos pos, EnumFacing facing)
+        {
+            super(world, pos, facing);
+        }
+    }
+    
+    /***
+     * Posts when a piston tries to retract
+     * If canceled the piston will not pull the block it is attached to
```
Comment (mezz) [inline src/main/java/net/minecraftforge/event/world/PistonEvent.java]: This javadoc is confusing. 
Is this only called for sticky pistons? 
What happens if it is canceled on regular pistons? 
Can retraction be completely canceled like extension can?

File: src/main/java/net/minecraftforge/event/ForgeEventFactory.java [hunk 2]
```diff
@@ -635,5 +635,13 @@ public static boolean canCreateFluidSource(World world, BlockPos pos, IBlockStat
         Result result = evt.getResult();
         return result == Result.DEFAULT ? def : result == Result.ALLOW;
     }
+    
+    public static boolean onPistonMove(World world, BlockPos pos, EnumFacing facing, boolean extending)
+    {
+        if (extending)
+            return MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.PistonEvent.PistonExtendEvent(world, pos, facing));
+        else
+            return MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.PistonEvent.PistonRetractEvent(world, pos, facing));
```
Comment (mezz) [inline src/main/java/net/minecraftforge/event/ForgeEventFactory.java]: Please import `PistonEvent`, this is not a patch so you can import anything you need to.

File: src/main/java/net/minecraftforge/event/ForgeEventFactory.java [hunk 3]
```diff
@@ -635,5 +635,13 @@ public static boolean canCreateFluidSource(World world, BlockPos pos, IBlockStat
         Result result = evt.getResult();
         return result == Result.DEFAULT ? def : result == Result.ALLOW;
     }
+    
+    public static boolean onPistonMove(World world, BlockPos pos, EnumFacing facing, boolean extending)
+    {
+        if (extending)
+            return MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.PistonEvent.PistonExtendEvent(world, pos, facing));
```
Comment (mezz) [inline src/main/java/net/minecraftforge/event/ForgeEventFactory.java]: Please import `PistonEvent`, this is not a patch so you can import anything you need to.

**Review by mezz — CHANGES_REQUESTED** [review]

**Comment by Selim042** [comment]
How do I mark that the requested changes have been taken care of?

**Review by mezz — APPROVED** [review]

File: src/test/java/net/minecraftforge/debug/EventPistonDebug.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+import net.minecraftforge.common.MinecraftForge;
+import net.minecraftforge.event.world.PistonEvent;
+import net.minecraftforge.fml.common.Mod;
+import net.minecraftforge.fml.common.event.FMLInitializationEvent;
+import net.minecraftforge.fml.common.event.FMLPreInitializationEvent;
+import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
+import net.minecraftforge.fml.common.registry.GameRegistry;
+
+@Mod(modid = EventPistonDebug.MODID)
+public class EventPistonDebug {
```
Comment (mezz) [inline src/test/java/net/minecraftforge/debug/EventPistonDebug.java]: Please fix the formatting of this file, all braces should be on a new line.

**Review by mezz — CHANGES_REQUESTED** [review]
Looks good, just needs some more formatting fixes and then I will test it.

**Comment by Selim042** [comment]
Did I miss formatting stuff or did you not get around to approving the changes?  Sorry, I'm just impatient sometimes.

File: src/main/java/net/minecraftforge/event/world/PistonEvent.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    private final EnumFacing facing;
+
+    public PistonEvent(World world, BlockPos pos, EnumFacing facing)
+    {
+        super(world, pos, world.getBlockState(pos));
+        
+        this.facing = facing;
+    }
+    
+    public EnumFacing getFacing() {
```
Comment (mezz) [inline src/main/java/net/minecraftforge/event/world/PistonEvent.java]: needs brace on new line

File: src/test/java/net/minecraftforge/debug/EventPistonDebug.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+                    new BlockPistonStructureHelper(event.getWorld(), event.getPos(), event.getFacing(), true);
+            for (EntityPlayer player : event.getWorld().playerEntities)
+            {
+                if (pistonHelper.canMove())
+                    player.addChatMessage(new TextComponentString(String.format("Piston will extend moving %d blocks and destroy %d blocks",
+                            pistonHelper.getBlocksToMove().size(), pistonHelper.getBlocksToDestroy().size())));
+                else
+                    player.addChatMessage(new TextComponentString("Piston won't extend"));
+            }
+        } else {
```
Comment (mezz) [inline src/test/java/net/minecraftforge/debug/EventPistonDebug.java]: needs braces on new lines
```
}
else
{
```

File: src/test/java/net/minecraftforge/debug/EventPistonDebug.java [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+                if (isSticky)
+                {
+                    BlockPistonStructureHelper pistonHelper =
+                            new BlockPistonStructureHelper(event.getWorld(), event.getPos(), event.getFacing(), false);
+                    if (pistonHelper.canMove())
+                        player.addChatMessage(new TextComponentString(String.format("Piston will retract moving %d blocks and destroy %d blocks",
+                                pistonHelper.getBlocksToMove().size(), pistonHelper.getBlocksToDestroy().size())));
+                    else
+                        player.addChatMessage(new TextComponentString("Piston won't retract"));
+                } else {
```
Comment (mezz) [inline src/test/java/net/minecraftforge/debug/EventPistonDebug.java]: braces on new lines

**Review by mezz — CHANGES_REQUESTED** [review]
Formatting looks mostly good. There are no notifications for when you update the code, if I'm not around you have to let me know by commenting on the issue.
I'll test it out now and can approve if the formatting is fixed and everything works well.

**Comment by mezz** [comment]
Also this should be updated to target the `1.11.x` branch. 
Forge is no longer accepting new events for the `1.10.x` branch.

**Comment by Selim042** [comment]
Fixed formatting, closing this to target 1.11.  I was hoping to have this in 1.10.x sadly.

**Comment by mezz** [comment]
Forge moves slow, sorry. I'm working to help speed up the PR process so that this is less of an issue in the future, but Forge will almost always be targeting the latest Minecraft version for features.

**Comment by liach** [comment]
@Selim042 For a future note, you can edit the target branch of a pull request now. No need for a new pull really.

**Comment by Selim042** [comment]
Oh.  Thanks!

**Pull request closed without merging**